firebase-firestore/src/main/java/com/google/firebase/firestore/util/CustomClassMapper.java

```diff
@@ -324,6 +334,19 @@ public static Object convertToPlainJavaTypes(Object object) {
       String value = (String) object;
       // We cast to Class without generics here since we can't prove the bound
       // T extends Enum<T> statically
+
+      // try to use PropertyName if exist
+      Field[] enumFields = clazz.getFields();
+      for (Field field : enumFields) {
+        if (field.isAnnotationPresent(PropertyName.class)) {
```

Comment:
This should reuse `propertyName` (or at least `annotationName`).

Also, not all fields in an enum class are necessarily enum constants, so I think you need to check `Field.isEnumConstant`. That would make the inner part of this loop look like this:

```java
if (field.isEnumConstant()) {
  String propertyName = propertyName(field);
  if (value.equals(propertyName)) {
    value = field.getName();
    break;
  }
}
```